### PR TITLE
Component Model

### PR DESCRIPTION
**Pull request by robertgraeff**
**How to categorize this PR?**

/area oci
/kind task
/priority 3

**What this PR does / why we need it**:

This pull request hides all read operations from an OCI registry, helm chart repository etc. behind a facade.
The purpose of the facade is to simplify the integration of OCM in the Landscaper. 

For example, the following operations will be "behind" the facade:

- reading a component descriptor and blueprint in the installation controller;
- reading a helm chart from an OCI registry or helm chart repository in the helm deployer;
- reading a JSON schema during the templating of deploy items;
- reading the component descriptor in the container deployer.

The facade is contained in package `pkg/components/model`.
The current implementation is in package `pkg/components/cnudie`.
Later another OCM-based implementation can be provided.

Package `pkg/components/registries` contains another implementation of the facade. It should make the switch between the cnudie and ocm implementations, i.e. it just delegates to one of the "real" implementations.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature developer
- a facade has been introduced that encapsulates the accesses to OCI registries and Helm chart repositories 
```

### CODE REVIEW

**Comment by robertgraeff** [comment]
Integration test has not yet  run.

File: pkg/components/testutils/componentversion.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+	cdv2 "github.com/gardener/component-spec/bindings-go/apis/v2"
+
+	lsv1alpha1 "github.com/gardener/landscaper/apis/core/v1alpha1"
+	"github.com/gardener/landscaper/pkg/components/model"
+	"github.com/gardener/landscaper/pkg/components/model/types"
+	"github.com/gardener/landscaper/pkg/landscaper/registry/componentoverwrites"
+)
+
+type TestComponentVersion struct {
```
Comment (achimweigel) [inline pkg/components/testutils/componentversion.go:17]: Why do we need this type? Isn't it the same as the one in .../cnudie/componentversion.go

File: pkg/landscaper/installations/builder.go [hunk 2]
```diff
@@ -141,44 +130,31 @@ func (b *OperationBuilder) Build(ctx context.Context) (*Operation, error) {
...
-			var err error
-			_, instOp.BlobResolver, err = instOp.ComponentsRegistry().
-				ResolveWithBlobResolver(ctx, cdRef.RepositoryContext, cdRef.ComponentName, cdRef.Version)
-			if err != nil {
-				return nil, err
-			}
+
+		componentVersion, err := registryAccess.GetComponentVersion(ctx, cdRef)
+		if err != nil {
+			return nil, err
```
Comment (achimweigel) [inline pkg/landscaper/installations/builder.go:142]: Looks much better but I fear I could not really verify this :-)

File: pkg/landscaper/registry/components/oci.go [hunk 3]
```diff
@@ -8,31 +8,16 @@ import (
...
-	if err != nil {
-		return nil, err
-	}
-
-	return NewOCIRegistryWithOCIClient(log, client, predefinedComponentDescriptors...)
-}
-
 // NewOCIRegistryWithOCIClient creates a new oci registry with a oci ociClient
 // If supplied, it parses and stores predefined component descriptors.
 func NewOCIRegistryWithOCIClient(log logging.Logger, client ociclient.Client, predefinedComponentDescriptors ...*cdv2.ComponentDescriptor) (TypedRegistry, error) {
```
Comment (achimweigel) [inline pkg/landscaper/registry/components/oci.go:23]: I'm wondering why NewOCIRegistryWithOCIClient is not in the components/cnudie package?

**Review by achimweigel — CHANGES_REQUESTED** [review]
really cool stuff - only some minor points (but very hard to review)

**Review by achimweigel — APPROVED** [review]
/lgtm

**Review by achimweigel — APPROVED** [review]
/lgtm

**Pull request merged**